### DISCUSSION TITLE
Looping Embla Carousel inside a container without overflow-hidden

### DISCUSSION THREAD

**Discussion by tobimori** [davidjerleke/embla-carousel#579] [Help]
I created a minimum reproducible example here: https://codepen.io/tobimori/pen/rNoddvj

Many designs are grid-based and have a container, and the contents of a slider are meant to be aligned to the container. When now using Embla's loop mode, embla uses the container measurements to decide whether to move items or not. This results in the loop being buggy/flickering. It works fine when not having the container, but then the alignment is off or you can't use a grid anymore.

It'd be great if we can supply a different container for looping measurements (in this case it would be body etc. instead). Or am I missing something and Embla already solves this?

The example above works just fine when removing the container class, but then the alignment is "off". Especially for grid based designs, where it's not possible to work around.

**Comment by davidjerleke** [comment]
Hi @tobimori,

Thank you for your question. 

> When now using Embla's loop mode, embla uses the container measurements to decide whether to move items or not. This results in the loop being buggy/flickering. It works fine when not having the container, but then the alignment is off or you can't use a grid anymore.

I think the docs [make it quite clear](https://www.embla-carousel.com/get-started/module/#the-html-structure) that the overflow wrapper **isn't optional**? So I'm not following the logic that Embla is buggy/flickering because you exclude markup that's necessary for the carousel to work. When looping the carousel, Embla somehow has to hide when slides shift position in order to create the loop effect. 

Align is working just fine, but you'll have to align with the [align option](https://www.embla-carousel.com/api/options/#align) and not with CSS.

> It'd be great if we can supply a different container for looping measurements (in this case it would be body etc. instead). Or am I missing something and Embla already solves this?

You can set your container and viewport to any size you want building upon [this guide](https://www.embla-carousel.com/get-started/module/#the-html-structure). However, you also need enough slides inside that container to loop and create the loop effect. Otherwise, ss [the docs say](https://www.embla-carousel.com/api/options/#loop):

> Embla automatically falls back to `false` if slide content isn't enough to create the loop effect without visible glitches.

Best,
David

**Comment by tobimori** [comment]
Let me try to explain it differently.
I want to use the following HTML structure:

```html
<div class="body"> -> this has `overflow: hidden`, but is `width: 100%`
  <div class="embla container"> -> this is my embla node & centered container with max-width
    <div class="wrapper"> -> this is my slide container
      <div class="slide"></div>
    </div>
  </div>
</div>
```

However the loop mode only moves slides outside of the size-limiting container, so it looks buggy. I want Embla to use the parent element of that (the .body element) to determine whether to move slides.

It's a pretty common use case for me to have a structure like this and I implemented this on almost all sites I've worked with Embla so far.

**Reply by tobimori** [reply]
![CleanShot 2023-09-26 at 11 49 17@2x](https://github.com/davidjerleke/embla-carousel/assets/29142128/f7c1350a-c021-4b8c-9c41-c159c20182ac)
Here's a visualization of what I want to do.
Important is the red grid in the background. The alignment should be determined by that element.
And it totally works with Embla right now! It's just the loop mode that's using a wrong size to calculate whether to move entries or not.

**Reply by tobimori** [reply]
```html
<div class="body embla">
  <div class="container">
    <div class="wrapper">
      <div class="slide"></div>
    </div>
  </div>
</div>
```
I tried specifying `container: '.wrapper'` and setting `emblaNode` to `.body`, but apparently embla always uses the direct parent of the slide container for size calculations and not the `emblaNode`?

**Reply by davidjerleke** [reply]
Hi @tobimori,

**On a side note**
CodeSandbox is more flexible than CodePen because I need an account to modify your pen. With CodeSandbox, I can just clone yours, modify it and send you the new link. No need for an account.

 in your CodePen, remove all your CSS and paste this instead:

```css
.container {
  width: 100%;
  margin: 0 auto;
  overflow: hidden;
}

.wrapper {
  display: flex;
}

.child {
  flex: 0 0 285px;
  height: 285px;
  background: #000;
  margin-right: 20px;
}
```

...and your JS with the following:

```js
const embla = EmblaCarousel(document.querySelector('.embla'), {
  loop: true,
  align: 'center',
  skipSnaps: true,
  slidesToScroll: 'auto'
})
```

Now to achieve **exactly** what you want you can statically set [`slidesToScroll`](https://www.embla-carousel.com/api/options/#slidestoscroll) to `4` instead of `auto`. But if you do that, you'll have to remember set different values for different breakpoints using the [`breakpoints`](https://www.embla-carousel.com/api/options/#breakpoints) option. Because using an absolute number instead of `'auto'` won't take the viewport width into consideration. It will just blindly follow.

Best,
David

**Reply by tobimori** [reply]
Hm, that's not quite what I want. With this, I don't get the alignment from my `.container` (aligned to 'start'), and I also want to scroll them one by one.
I made a modified CodeSandbox that better shows what I want: https://codesandbox.io/s/embla-carousel-generator-vanilla-forked-3hg4nq

![CleanShot 2023-09-26 at 14 27 39@2x](https://github.com/davidjerleke/embla-carousel/assets/29142128/fc9e3a67-e8a6-40c8-babc-d0d313078556)

**Reply by davidjerleke** [reply]
@tobimori now you changed the width of your items so it's a bit confusing. If you want to scroll one slide at a time you can remove the `slidesToScroll` option. You can set `align` to a **custom number** instead of the predefined string. From [the docs](https://www.embla-carousel.com/api/options/#align):

> Align the slides relative to the carousel viewport. Use one of the predefined alignments `start`, `center` or `end`. Alternatively, provide a number between `0` - `1` to align the slides, where **0.5 equals 50%**. 

Best,
David

**Reply by tobimori** [reply]
I changed the width so you can better see what I want to do with the container. So the solution would be to manually calculate the spacing the container has on each side?

**Reply by davidjerleke** [reply]
@tobimori in this case yes. I think that's the way you want to go because you need the container to be 100% wide to cover the loop effect properly and you want to align it to the left (start) of an imaginary container of `max-width: 1200px`.

**Reply by davidjerleke** [reply]
@tobimori we could **replace the number option** with a **callback** function like so:

```ts
const options = {
  align: (viewSize, snapSize, index) => {
    // create your own alignment
    return yourCustomAlignment
  }
}
```

The benefit of this would be that you would have direct access to the size of the container (viewSize) and the current slide/slideGroup (itemSize) which is convenient because you don't have to read element dimensions yourself. Note that the **function** would **run** for **each and every scroll snap**. This would be even more flexible than the previous implementation and wouldn't affect bundle size.

Let me know if it's of interest.

Best,
David

**Reply by davidjerleke** [reply]
@tobimori closing this if you're not [interested in this](https://github.com/davidjerleke/embla-carousel/discussions/579#discussioncomment-7120106).

**Reply by tobimori** [reply]
Sorry, I forgot to respond, but this would be extremely helpful! 😍

**Reply by davidjerleke** [reply]
- Issue created from this discussion #594.

**Reply by davidjerleke** [reply]
@tobimori I've released this feature in [`v8.0.0-rc15`](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.0.0-rc15).
